docs/style_guide.md

```diff
@@ -0,0 +1,159 @@
+making it easier to understand how the function affects the surrounding code.
+
+### Inline constants
+
+If a constant is private and used only once, prefer inlining it where it's used
+instead of adding a top-level const declaration.
+
+Regexes are exempt from this since they should be initialized at startup - See
+the section below for more details.
+
+### Avoid init()
+
+Usage of `init()` makes it hard to keep track of the control flow. Prefer
+avoiding it in production code. In tests you can use `TestMain()` instead.
+
+### Context propagation
+
+SCALIBR library users can pass in a `context.Context` to control timeouts and
+context cancellation. Make sure this top-level context is passed down to lower
+functions (don't initialize a new context with `context.Background()`) and check
+for context cancellation whenever something long-running is performed such as
+looping ([example](https://github.com/google/osv-scalibr/blob/8b03d0859edf445152f34c420f50ffe0abf057df/extractor/filesystem/os/dpkg/dpkg.go#L183)).
+
+## Error handling
+
+### Avoid panics
+
+If a plugin encounters an error the rest of SCALIBR and the callers' code
+shouldn't crash. Avoid calling panics and prefer propagating errors instead.
+
+### Init regexes at startup time
+
+Add all regex definitions that use `MustCompile` as global vars that initialize
+at startup time ([example](https://github.com/google/osv-scalibr/blob/8b03d0859edf445152f34c420f50ffe0abf057df/extractor/filesystem/os/nix/nix.go#L92)). This allows the initialization computation to be done up front and
+catches any potential crashes before the scan runs.
+
+### Propagate or log errors
+
+In general, propagate errors upwards to the caller.
+
+If the error is expected or not something that should make the module fail
+(e.g. an Extractor encountered an invalid package.json file) there's no need to
+propagate it but consider logging a warning instead.
+
+## Testing
+
+### Avoid assertion libraries
+
+Generally avoid creating helper libraries that [perform test assertions](https://google.github.io/styleguide/go/decisions.html#assertion-libraries).
+Instead, use helper libs to transform your data into a more easily comparable
+structure and perform the comparisions/assertions in the main test function.
+Example: the [extracttest](https://github.com/google/osv-scalibr/blob/8b03d0859edf445152f34c420f50ffe0abf057df/extractor/filesystem/language/dart/pubspec/pubspec_test.go#L296)
+helper lib.
+
+An exception is when the helper library is used to set up the testing
+environment (e.g. create specific files). In these cases it's fine to assert
+that the setup succeeded in the library function as long as the setup code is
+not related to the functionality being tested ([example](https://github.com/google/osv-scalibr/blob/8b03d0859edf445152f34c420f50ffe0abf057df/extractor/filesystem/os/dpkg/dpkg_test.go#L1527)).
+
+### Use easy to find subtest descriptions
```

Comment:
@erikvarga @another-rex am I understanding that right that we should be doing e.g. "v1_layer_missing_diffID" for the name, rather than "v1 layer missing diffID"? And if so, do you want a hand converting the existing subtests?

(likewise, do we want to do this over in `osv-scanner` too?)

Reply:
Yes

I've avoided adding this to osv-scanner because of the work required to convert all the tests, but ideally we have this in osv-scanner as well. Being able to Ctrl f a test would be really nice. (I feel like there should be a VSCode extension to ctrl click a test to jump to that table test case).

If you can make a quick script to replace all the space in tests with underscores that'll be great.